tests/bsim/bluetooth/host/l2cap/reassembly/peer/src/peer.c

```diff
@@ -0,0 +1,650 @@
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/__assert.h>
+
+#include <zephyr/net/buf.h>
+#include <zephyr/bluetooth/buf.h>
+
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/hci.h>
+#include <zephyr/bluetooth/hci_raw.h>
+
+#include "common/bt_str.h"
+
+#include "host/conn_internal.h"
+#include "host/l2cap_internal.h"
+
+#include "babblekit/flags.h"
+#include "babblekit/testcase.h"
+
+#include "data.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_tinyhost, LOG_LEVEL_INF);
+
+#define BT_ATT_OP_MTU_REQ   0x02
+#define BT_ATT_OP_MTU_RSP   0x03
+#define BT_ATT_OP_WRITE_REQ 0x12
+#define BT_ATT_OP_WRITE_RSP 0x13
+#define BT_ATT_OP_NOTIFY    0x1b
+#define BT_ATT_OP_INDICATE  0x1d
+#define BT_ATT_OP_CONFIRM   0x1e
+#define BT_ATT_OP_WRITE_CMD 0x52
+#define BT_L2CAP_CID_ATT    0x0004
+#define LAST_SUPPORTED_ATT_OPCODE 0x20
+
+DEFINE_FLAG(is_connected);
+DEFINE_FLAG(flag_data_length_updated);
+DEFINE_FLAG(flag_handle);
+DEFINE_FLAG(flag_notified);
+DEFINE_FLAG(flag_write_ack);
+DEFINE_FLAG(flag_req_in_progress);
+
+static K_FIFO_DEFINE(rx_queue);
+
+#define CMD_BUF_SIZE MAX(BT_BUF_EVT_RX_SIZE, BT_BUF_CMD_TX_SIZE)
+NET_BUF_POOL_FIXED_DEFINE(hci_cmd_pool, CONFIG_BT_BUF_CMD_TX_COUNT,
+			  CMD_BUF_SIZE, 8, NULL);
+
+static K_SEM_DEFINE(cmd_sem, 1, 1);
+static struct k_sem acl_pkts;
+static uint16_t conn_handle;
+
+static volatile uint16_t active_opcode = 0xFFFF;
+static struct net_buf *cmd_rsp;
+
+struct net_buf *bt_hci_cmd_create(uint16_t opcode, uint8_t param_len)
+{
+	struct bt_hci_cmd_hdr *hdr;
+	struct net_buf *buf;
+
+	LOG_DBG("opcode 0x%04x param_len %u", opcode, param_len);
+
+	buf = net_buf_alloc(&hci_cmd_pool, K_FOREVER);
+	TEST_ASSERT(buf, "failed allocation");
+
+	LOG_DBG("buf %p", buf);
+
+	net_buf_reserve(buf, BT_BUF_RESERVE);
+
+	bt_buf_set_type(buf, BT_BUF_CMD);
+
+	hdr = net_buf_add(buf, sizeof(*hdr));
+	hdr->opcode = sys_cpu_to_le16(opcode);
+	hdr->param_len = param_len;
+
+	return buf;
+}
+
+static void handle_cmd_complete(struct net_buf *buf)
+{
+	struct bt_hci_evt_hdr *hdr;
+	uint8_t status, ncmd;
+	uint16_t opcode;
+	struct net_buf_simple_state state;
+
+	net_buf_simple_save(&buf->b, &state);
+
+	hdr = net_buf_pull_mem(buf, sizeof(*hdr));
+
+	if (hdr->evt == BT_HCI_EVT_CMD_COMPLETE) {
+		struct bt_hci_evt_cmd_complete *evt;
+
+		evt = net_buf_pull_mem(buf, sizeof(*evt));
+		status = 0;
+		ncmd = evt->ncmd;
+		opcode = sys_le16_to_cpu(evt->opcode);
+
+	} else if (hdr->evt == BT_HCI_EVT_CMD_STATUS) {
+		struct bt_hci_evt_cmd_status *evt;
+
+		evt = net_buf_pull_mem(buf, sizeof(*evt));
+		status = buf->data[0];
+		ncmd = evt->ncmd;
+		opcode = sys_le16_to_cpu(evt->opcode);
+
+	} else {
+		TEST_FAIL("unhandled event 0x%x", hdr->evt);
+	}
+
+	LOG_DBG("opcode 0x%04x status %x", opcode, status);
+
+	TEST_ASSERT(status == 0x00, "cmd 0x%x status: 0x%x", opcode, status);
+
+	TEST_ASSERT(active_opcode == opcode, "unexpected opcode %x != %x", active_opcode, opcode);
+
+	if (active_opcode) {
```

Comment:
Looks like `active_opcode` will always be true as we set it to `0xFFFF` or some other opcode. So I suggest removing this conditional statement.